$OpenBSD: patch-Source_cmSystemTools_cxx,v 1.16 2020/12/30 07:17:34 rsadowski Exp $

Get rid of cmLocaleRAII.

It is of dubious utility:

it's trying to save the LC_CTYPE locale and set it from the environment
variables before processing each single file name, and to restore
the LC_CTYPE locale to the previous value after processing each
file name. There is no way how the environment could possibly change
between files, and setlocale(3) is a potentially expensive operation.

The way how it tries to save the previous locale is broken:

it calls setlocale(LC_CTYPE, NULL), which returns a pointer. CMake
saves that pointer, without strdup(3)ing it, then calls setlocale(3)
again as setlocale(LC_CTYPE, "") to actually set the desired locale,
discarding the return value. It later uses the saved pointer to try
and restore the original locale.
In an implementation using a static buffer, the second call will
overwrite the contents of the pointer returned by the first call,
so the third call will not restore the previous locale, but instead
set the one intended as temporary a second time.
A stricter implementation (one that doesn't just overwrite the
string, but actually frees the memory, so the string returned by
the earlier setlocale() call becomes invalid) will crash, revealing
the invalid use of the interface.

Index: Source/cmSystemTools.cxx
--- Source/cmSystemTools.cxx.orig
+++ Source/cmSystemTools.cxx
@@ -1,11 +1,11 @@
 /* Distributed under the OSI-approved BSD 3-Clause License.  See accompanying
    file Copyright.txt or https://cmake.org/licensing for details.  */
 
-#if !defined(_WIN32) && !defined(__sun)
+#if !defined(_WIN32) && !defined(__sun) && !defined(__OpenBSD__)
 // POSIX APIs are needed
 #  define _POSIX_C_SOURCE 200809L
 #endif
-#if defined(__OpenBSD__) || defined(__FreeBSD__) || defined(__NetBSD__) ||    \
+#if defined(__FreeBSD__) || defined(__NetBSD__) ||    \
   defined(__QNX__)
 // For isascii
 #  define _XOPEN_SOURCE 700
@@ -1674,8 +1674,6 @@ bool extract_tar(const std::string& outFileName,
                  const std::vector<std::string>& files, bool verbose,
                  bool extract)
 {
-  cmLocaleRAII localeRAII;
-  static_cast<void>(localeRAII);
   struct archive* a = archive_read_new();
   struct archive* ext = archive_write_disk_new();
   archive_read_support_filter_all(a);
